config/config.go

```diff
@@ -446,8 +448,46 @@ type ScrapeConfig struct {
 
 	// List of target relabel configurations.
 	RelabelConfigs []*relabel.Config `yaml:"relabel_configs,omitempty"`
+
 	// List of metric relabel configurations.
 	MetricRelabelConfigs []*relabel.Config `yaml:"metric_relabel_configs,omitempty"`
+
+	// List of rules to execute at scrape time.
+	RuleConfigs []*ScrapeRuleConfig `yaml:"scrape_rule_configs,omitempty"`
+}
+
+// ScrapeRuleConfig is the configuration for rules executed
+// at scrape time for each individual target.
+type ScrapeRuleConfig struct {
+	Expr   string `yaml:"expr"`
+	Record string `yaml:"record"`
+}
+
+func (a *ScrapeRuleConfig) Validate() error {
+	if a.Record == "" {
+		return errors.New("aggregation rule record must not be empty")
+	}
+
+	if a.Expr == "" {
+		return errors.New("aggregation rule expression must not be empty")
+	}
+
+	expr, err := parser.ParseExpr(a.Expr)
+	if err != nil {
+		return errors.Wrapf(err, "invalid scrape rule expression")
+	}
+
+	parser.Inspect(expr, func(node parser.Node, nodes []parser.Node) error {
+		switch node.(type) {
+		case *parser.MatrixSelector:
+			err = errors.New("matrix selectors are not allowed in scrape rule expressions")
```

Comment:
Historically there was always the idea that some day scrape-time rules could allow at least instantaneous rate (`irate()`) calculations, which would require looking back one more sample at least. You could say that dimensional aggregation in scrape-time rules is much more important than that, but the problem is that a lot of metrics are counters, which can't be aggregated (or used in other useful ways) before being rate-ed.

However, now that I think about it more, counters within a *single* target would almost always reset together at the same time (unless you are exposing state for multiple third parties keeping counter states on their own), so maybe it would allow you to get around the "rate-before-sum" rule in this special case.

Reply:
The main problem there is a counter being removed, so rate-before-sum still applies. To solve it properly, you need to auto-add the target labels to the selector and fetch the data from the tsdb.

Reply:
Another case where a counter could reset are overflows. Would that also cause problems with sum before rate?

Reply:
It would, you can only violate rate-before-sum if you know exactly what you're doing and it thus can never be general advice.

Reply:
I also wonder if fetching data from tsdb can be avoided by keeping samples from previous scrapes cached in each scrape loop. If we want to use samples from tsdb, then we can't drop series after aggregation and we still have to pay for having them in tsdb. This kind of defeats the purpose of aggregating at scrape time and I am not sure if it is much better than having a regular recording rule.

Reply:
The purpose of scrape-time evaluation is to spread CPU load around, so the amount of data a recording rule can handle is less of a limiting factor. Recording rules are (basically) single threaded, whereas many scrapes can happen concurrently.
That the results will also have the same timestamp as the scrape can also be handy.

Reply:
One common want in my org is to have a per cluster aggregate without the need to store all the per instance metrics. So my initial hope was for that - build an aggregate from scrape results, save it and discard all those metrics. But because scrapes are per target it won't work.
If it was possible then it would be a big win for some services we run since it's just too expensive to scrape all per instance metrics just so that you can build an aggregate.
That being said it's entirely possible that same could be achieved with ability to set per metric retention period since we could keep only enough samples for per instance metrics to allow aggregation rules to run.

Reply:
More thoughts about counter resets: We don't really get overflows with our float64 counters. Instead, we run into "loss of precision" areas, which could be taken as a reason to reset an individual counter. I haven't heard of anyone doing this yet, but it could be done. And, fun fact, the upcoming new histograms use a similar trick: Resetting the histogram to shed entropy once too many buckets have been populated.

I would suggest to keep matrix selectors out of the game for now and think about them in a later iteration of the feature.

Reply:
Right, that's where the counter comes from a different monitoring system.

Anyway, bottom line is: counter resets can happen for a whole lot of reasons, not only if the binary restarts. It's really hard to rule them out.